music21/metadata/__init__.py

```diff
@@ -308,48 +743,281 @@ def all(self, skipContributors=False):
+            if len(values) == 2:
+                return str(values[0]) + ' and ' + str(values[1])
+            return str(values[0]) + f' and {len(values)-1} others'
+
+        if self._needsArticleNormalization(nsKey):
+            output: str = ''
+            for i, value in enumerate(values):
+                assert isinstance(value, Text)
+                if i > 0:
+                    output += ', '
+                output += value.getNormalizedArticle()
+            return output
+
+        return ', '.join(str(value) for value in values)
+
+    def _getStringValuesByNSKey(self, nsKey: str) -> t.List[str]:
+        values: t.Tuple[ValueType, ...]
+        try:
+            values = self._get(nsKey, isCustom=False)
+        except KeyError:
+            return []
+
+        if not values:
+            return []
+
+        if self._needsArticleNormalization(nsKey):
+            output: t.List[str] = []
+            for value in values:
+                assert isinstance(value, Text)
+                output.append(value.getNormalizedArticle())
+            return output
+
+        return [str(value) for value in values]
+
+    def _getPluralAttribute(self, attributeName: str) -> t.List[str]:
+        # This does what __getattr__ would do if we supported plural attributeNames
+        # (but it takes singular attributeNames, of course).
+
+        # It raises AttributeError if attributeName is not a valid uniqueName,
+        # workId, or workId abbreviation.  Used in search, for example, because
+        # search wants to find everything.
+
+        if attributeName in properties.UNIQUE_NAME_TO_NSKEY:
+            return self._getStringValuesByNSKey(properties.UNIQUE_NAME_TO_NSKEY[attributeName])
+
+        # Is attributeName a grandfathered workId?
+        if attributeName in properties.MUSIC21_WORK_ID_TO_NSKEY:
+            return self._getStringValuesByNSKey(properties.MUSIC21_WORK_ID_TO_NSKEY[attributeName])
+
+        # Is attributeName a grandfathered workId abbreviation?
+        if attributeName in properties.MUSIC21_ABBREVIATION_TO_NSKEY:
+            return self._getStringValuesByNSKey(
+                properties.MUSIC21_ABBREVIATION_TO_NSKEY[attributeName])
+
+        # The following are in searchAttributes, and getattr will find them because
+        # they are a property, but this routine needs to find them, too.
+        if attributeName == 'fileFormat':
+            if self.fileFormat is None:
+                return []
+            return [self.fileFormat]
+
+        if attributeName == 'filePath':
+            if self.filePath is None:
+                return []
+            return [self.filePath]
+
+        if attributeName == 'fileNumber':
+            if self.fileNumber is None:
+                return []
+            return [str(self.fileNumber)]
+
+        raise AttributeError(f'invalid attributeName: {attributeName}')
+
+    def _getSingularAttribute(self, attributeName: str) -> t.Optional[str]:
```

Comment:
as above, docs + examples.

Reply:
I think the important thing here is "how easy is it to get it right in a parser, given what we write in the writer".  Our own parser obviously can get it right either way, but do other folks try to interpret the contents of <miscellaneous-field>?  If not, maybe they should, and having names from well-known namespaces would be a good start.  (I would also add that the way we put date values in MusicXML could be improved: it's hard to parse things like '1689/--/-- or earlier'.  Our own parser doesn't even do it yet (I think).  We should probably use some standard date string-ifying here.)